File: bunood_theme/public/js/asset_workbench.js
// Copyright (c) 2026, Bunood and contributors
// A bilingual work queue over ERPNext's native fixed-asset lifecycle.

/* eslint-env browser */
/* global frappe, __ */

(function () {
	"use strict";

	const METHOD = "bunood_theme.api.asset_workbench";

	function el(tag, className, text) {
		const node = document.createElement(tag);
		if (className) node.className = className;
		if (text !== undefined && text !== null) node.textContent = text;
		return node;
	}

	function actionButton(label, action, primary, enabled = true) {
		const node = el("button", `btn ${primary ? "btn-primary" : "btn-default"}`, label);
		node.type = "button";
		node.disabled = !enabled;
		if (enabled) node.addEventListener("click", action);
		return node;
	}

	function routeButton(label, route, primary, enabled = true) {
		return actionButton(label, () => frappe.set_route(...route), primary, enabled);
	}

	function newDocumentButton(label, doctype, data, enabled) {
		return actionButton(label, () => frappe.new_doc(doctype, data), false, enabled);
	}

	function control(host, options) {
		const field = frappe.ui.form.make_control({ parent: host, df: options, render_input: true });
		field.refresh();
		return field;
	}

	function formatAmount(value, currency) {
		try {
			return new Intl.NumberFormat(frappe.boot.lang || undefined, {
				style: "currency",
				currency: currency || "SAR",
				minimumFractionDigits: 2,
			}).format(Number(value || 0));
		} catch (_error) {
			return `${Number(value || 0).toFixed(2)} ${currency || ""}`.trim();
		}
	}

	function stateCopy(state) {
		return {
			"attention-needed": [__("Asset attention is required"), __("Review due or failed depreciation and open repairs in their native records before close.")],
			"review-required": [__("Asset review is required"), __("Draft asset lifecycle records still require authorised review and native submission.")],
			"incomplete-evidence": [__("Asset evidence is incomplete"), __("A permitted query failed or a bounded result was truncated. No clean-register conclusion was inferred.")],
			"no-exceptions-observed": [__("No exceptions observed in this bounded view"), __("This is not fixed-asset or General Ledger reconciliation. Complete the native reports and qualified review.")],
		}[state] || [__("Asset review is required"), __("Review the native asset records for this company and period.")];
	}

	function assetState(state) {
		return {
			"depreciation-failed": __("Depreciation posting failed"),
			"depreciation-due": __("Depreciation due by boundary"),
			"draft-review": __("Draft — review required"),
			"service-attention": __("Maintenance or repair attention"),
			"fully-depreciated": __("Fully depreciated"),
			"in-use": __("In use"),
			disposed: __("Disposed"),
			cancelled: __("Cancelled"),
		}[state] || __("Review required");
	}

	function documentState(docstatus, state) {
		if (Number(docstatus) === 2) return __("Cancelled");
		if (Number(docstatus) === 0) return __("Draft — review required");
		return state ? __(state) : __("Submitted");
	}

	function metric(label, value, detail, warning) {
		const card = el("article", `bnd-close__metric${warning ? " is-warning" : ""}`);
		card.append(el("span", "", label), el("strong", "", String(value)), el("small", "", detail));
		return card;
	}

	function renderLifecycle(data) {
		const panel = el("section", "bnd-close__panel");
		panel.append(
			el("h2", "", __("Fixed-asset lifecycle")),
			el("p", "bnd-close__note", __("Physical custody and accounting value are different controls. Use the native document that matches the event."))
		);
		const grid = el("div", "bnd-close__controls");
		const choices = [
			[__("Acquire and capitalize"), __("Create the asset from its fixed-asset Item or combine eligible costs through Asset Capitalization."), __("New capitalization"), "Asset Capitalization", data.capabilities.can_create_capitalization],
			[__("Put in use and depreciate"), __("Confirm available-for-use date, Finance Book, method, useful life, residual value and schedule before posting."), __("Depreciation schedules"), null, true, ["List", "Asset Depreciation Schedule"]],
			[__("Move and assign custody"), __("Use Asset Movement for location and custodian changes. Do not edit the visible location as an unaudited shortcut."), __("New movement"), "Asset Movement", data.capabilities.can_create_movement],
			[__("Maintain and repair"), __("Keep planned maintenance separate from failures, repair cost and any policy-approved capitalization."), __("New repair"), "Asset Repair", data.capabilities.can_create_repair],
			[__("Adjust or dispose"), __("Use a supported value adjustment, fixed-asset sale or scrap workflow so carrying value and gain or loss remain traceable."), __("Value adjustment"), "Asset Value Adjustment", data.capabilities.can_create_adjustment],
		];
		choices.forEach(([title, note, label, doctype, enabled, route]) => {
			const card = el("article", "bnd-close__control");
			const button = route ? routeButton(label, route, false) : newDocumentButton(label, doctype, { company: data.company }, enabled);
			card.append(el("h3", "", title), el("p", "", note), button);
			grid.append(card);
		});
		panel.append(grid);
		return panel;
	}

	function renderAssetQueue(data) {
		const panel = el("section", "bnd-close__panel");
		panel.append(
			el("h2", "", __("Current asset queue")),
			el("p", "bnd-close__note", __("The period filters lifecycle activity. This queue shows the current permitted Asset register state, prioritised by attention."))
		);
		if (!data.assets.length) {
			panel.append(el("p", "bnd-close__empty", __("No permitted Asset records were observed for this company.")));
			return panel;
		}
		const wrap = el("div", "bnd-close__table-wrap");
		const table = el("table", "bnd-close__table");
		const head = el("tr");
		[__("Asset"), __("Category"), __("Lifecycle state"), __("Location or custodian"), __("Next depreciation"), __("Carrying value")].forEach((label) => head.append(el("th", "", label)));
		const thead = el("thead");
		thead.append(head);
		const body = el("tbody");
		data.assets.forEach((row) => {
			const tr = el("tr");
			const nameCell = el("td");
			const link = el("button", "bnd-close__link", row.asset_name || row.name);
			link.type = "button";
			link.addEventListener("click", () => frappe.set_route("Form", "Asset", row.name));
			nameCell.append(link, el("small", "", row.item_code ? ` · ${row.item_code}` : ""));
			const custody = [row.location, row.custodian].filter(Boolean).join(" · ") || "—";
			tr.append(
				nameCell,
				el("td", "", row.asset_category || "—"),
				el("td", "", assetState(row.state)),
				el("td", "", custody),
				el("td", "", row.next_depreciation_date || "—"),
				el("td", "", formatAmount(row.carrying_value, data.currency))
			);
			body.append(tr);
		});
		table.append(thead, body);
		wrap.append(table);
		panel.append(wrap);
		if (data.asset_display_truncated || data.evidence_truncated) {
			panel.append(el("p", "bnd-close__note", __("This is a bounded queue. Open the native Asset list and reports for the complete permitted result.")));
		}
		return panel;
	}

	function renderSchedules(data) {
		const panel = el("section", "bnd-close__panel");
		panel.append(
			el("h2", "", __("Depreciation schedules")),
			el("p", "bnd-close__note", __("A schedule is not a posted expense. Verify each Finance Book, future row and linked Journal Entry in the native schedule and ledger report."))
		);
		if (!data.schedules.length) {
			panel.append(el("p", "bnd-close__empty", __("No permitted Asset Depreciation Schedule was observed.")));
			return panel;
		}
		const wrap = el("div", "bnd-close__table-wrap");
		const table = el("table", "bnd-close__table");
		const head = el("tr");
		[__("Schedule"), __("Asset"), __("Finance Book"), __("Method"), __("State"), __("Current scheduled value")].forEach((label) => head.append(el("th", "", label)));
		const thead = el("thead");
		thead.append(head);
		const body = el("tbody");
		data.schedules.forEach((row) => {
			const tr = el("tr");
			const nameCell = el("td");
			const link = el("button", "bnd-close__link", row.name);
			link.type = "button";
			link.addEventListener("click", () => frappe.set_route("Form", "Asset Depreciation Schedule", row.name));
			nameCell.append(link);
			tr.append(nameCell, el("td", "", row.asset || "—"), el("td", "", row.finance_book || data.finance_book || "—"), el("td", "", __(row.depreciation_method || "—")), el("td", "", documentState(row.docstatus, row.status)), el("td", "", formatAmount(row.value_after_depreciation, data.currency)));
			body.append(tr);
		});
		table.append(thead, body);
		wrap.append(table);
		panel.append(wrap);
		return panel;
	}

	function renderActivity(data) {
		const panel = el("section", "bnd-close__panel");
		panel.append(
			el("h2", "", __("Lifecycle activity and open repairs")),
			el("p", "bnd-close__note", __("Open the source document to verify affected assets, stock or service cost, accounting date, evidence and approval history."))
		);
		if (!data.activity.length) {
			panel.append(el("p", "bnd-close__empty", __("No permitted lifecycle activity or open repair was observed in this boundary.")));
			return panel;
		}
		const wrap = el("div", "bnd-close__table-wrap");
		const table = el("table", "bnd-close__table");
		const head = el("tr");
		[__("Document"), __("Type"), __("Date"), __("Asset or purpose"), __("State"), __("Amount")].forEach((label) => head.append(el("th", "", label)));
		const thead = el("thead");
		thead.append(head);
		const body = el("tbody");
		data.activity.forEach((row) => {
			const tr = el("tr");
			const nameCell = el("td");
			const link = el("button", "bnd-close__link", row.name);
			link.type = "button";
			link.addEventListener("click", () => frappe.set_route("Form", row.doctype, row.name));
			nameCell.append(link);
			tr.append(nameCell, el("td", "", __(row.doctype)), el("td", "", row.date || "—"), el("td", "", row.asset || row.title || "—"), el("td", "", documentState(row.docstatus, row.state)), el("td", "", formatAmount(row.amount, data.currency)));
			body.append(tr);
		});
		table.append(thead, body);
		wrap.append(table);
		panel.append(wrap);
		return panel;
	}

	function renderReconciliation(data) {
		const panel = el("section", "bnd-close__panel");
		panel.append(
			el("h2", "", __("Register-to-ledger review")),
			el("p", "bnd-close__note", __("Use the same company, cutoff date, Finance Book, currency and dimensions. Reconcile asset cost, accumulated depreciation, depreciation expense, value adjustments and disposal gain or loss; investigate manual journals and timing differences."))
		);
		const actions = el("div", "bnd-close__report-grid");
		[
			[__("Fixed Asset Register"), ["query-report", "Fixed Asset Register"], data.capabilities.can_read_asset],
			[__("Depreciation and balances"), ["query-report", "Asset Depreciations and Balances"], data.capabilities.can_read_asset],
			[__("Asset Depreciation Ledger"), ["query-report", "Asset Depreciation Ledger"], data.capabilities.can_read_gl],
			[__("Asset Activity"), ["query-report", "Asset Activity"], data.capabilities.can_read_asset],
			[__("General Ledger"), ["query-report", "General Ledger"], data.capabilities.can_read_gl],
			[__("Finance and close"), ["bnd-finance-close"], true],
		].forEach(([label, route, enabled]) => actions.append(routeButton(label, route, false, enabled)));
		panel.append(actions, el("p", "bnd-close__note", __("Bunood has not declared this boundary reconciled. Qualified preparer and reviewer evidence remains required.")));
		return panel;
	}

	function renderResult(data, content) {
		content.replaceChildren();
		const state = el("section", `bnd-close__state is-${data.state}`);
		const copy = stateCopy(data.state);
		state.setAttribute("role", "status");
		state.setAttribute("aria-live", "polite");
		state.append(el("div", "", copy[0]), el("p", "", copy[1]), el("small", "", `${data.company} · ${data.from_date} — ${data.to_date} · ${__("Generated")} ${data.generated_at}`));

		const attention = data.summary.depreciation_due + data.summary.depreciation_failed + data.summary.pending_repairs;
		const metrics = el("section", "bnd-close__metrics");
		metrics.append(
			metric(__("Assets observed"), data.summary.observed_assets, __("Current permitted register rows"), data.evidence_truncated),
			metric(__("Attention queue"), attention, __("Due or failed depreciation and open repairs"), attention > 0),
			metric(__("Observed carrying value"), formatAmount(data.summary.observed_carrying_value, data.currency), __("Not a reconciled register total"), data.evidence_truncated),
			metric(__("Period activity"), data.summary.period_activity + data.summary.disposed_in_period, __("Movements, capitalization, adjustments and disposal"), false)
		);

		content.append(state, metrics, renderLifecycle(data), renderAssetQueue(data), renderSchedules(data), renderActivity(data), renderReconciliation(data));
	}

	function render(container) {
		container.replaceChildren();
		const root = el("main", "bnd-close bnd-assets");
		const intro = el("header", "bnd-close__intro");
		const introCopy = el("div", "bnd-close__intro-copy");
		introCopy.append(
			el("span", "bnd-close__eyebrow", __("Native ERPNext asset lifecycle")),
			el("h1", "", __("Fixed asset workbench")),
			el("p", "", __("Track acquisition, custody, depreciation, maintenance, value changes and disposal without hiding native accounting controls."))
		);
		const actions = el("div", "bnd-close__actions");
		const newAsset = actionButton(__("New Asset"), () => frappe.new_doc("Asset", { company: company.get_value() }), true);
		newAsset.disabled = true;
		actions.append(newAsset, routeButton(__("Assets"), ["List", "Asset"], false));
		intro.append(introCopy, actions);

		const filters = el("section", "bnd-close__filters");
		const companyHost = el("div", "bnd-close__field");
		const fromHost = el("div", "bnd-close__field");
		const toHost = el("div", "bnd-close__field");
		const applyHost = el("div", "bnd-close__apply");
		const apply = el("button", "btn btn-primary", __("Review assets"));
		apply.type = "button";
		applyHost.append(apply);
		filters.append(companyHost, fromHost, toHost, applyHost);
		const content = el("div", "bnd-close__content");
		root.append(intro, filters, content);
		container.append(root);

		const company = control(companyHost, { fieldtype: "Link", fieldname: "company", label: __("Company"), options: "Company", reqd: 1 });
		const from = control(fromHost, { fieldtype: "Date", fieldname: "from_date", label: __("From date"), reqd: 1 });
		const to = control(toHost, { fieldtype: "Date", fieldname: "to_date", label: __("To date"), reqd: 1 });
		company.set_value(frappe.defaults.get_user_default("Company") || "");
		to.set_value(frappe.datetime.get_today());
		from.set_value(`${frappe.datetime.get_today().slice(0, 4)}-01-01`);

		const load = () => {
			if (!company.get_value()) {
				frappe.msgprint(__("Company is required"));
				return;
			}
			root.classList.add("is-loading");
			apply.disabled = true;
			newAsset.disabled = true;
			frappe.call({
				method: METHOD,
				args: { company: company.get_value(), from_date: from.get_value(), to_date: to.get_value() },
			}).then((response) => {
				renderResult(response.message, content);
				newAsset.disabled = !response.message.capabilities.can_create_asset;
			}).catch(() => {
				const failure = el("section", "bnd-close__failure");
				failure.setAttribute("role", "alert");
				failure.append(el("h2", "", __("The fixed-asset workbench could not be loaded")), el("p", "", __("No asset or accounting record was changed. Review the error, then try again.")));
				content.replaceChildren(failure);
			}).finally(() => {
				root.classList.remove("is-loading");
				apply.disabled = false;
			});
		};
		apply.addEventListener("click", load);
		if (company.get_value()) load();
	}

	window.bunood_theme = window.bunood_theme || {};
	window.bunood_theme.asset_workbench_render = render;
})();

File: bunood_theme/asset_workbench.py
# Copyright (c) 2026, Bunood and contributors
# For license information, please see license.txt
"""Permission-safe context for Bunood's native fixed-asset workbench.

The workbench organises ERPNext's native Asset and lifecycle documents. It
does not calculate depreciation, post to the General Ledger, move, capitalize,
repair, adjust, sell, scrap, approve, or maintain a parallel asset register.
"""

from __future__ import annotations

import frappe
from frappe.utils import date_diff, flt, getdate, now_datetime, nowdate


MAX_PERIOD_DAYS = 366
ASSET_QUERY_LIMIT = 1001
AUXILIARY_QUERY_LIMIT = 201
ASSET_DISPLAY_LIMIT = 100
ACTIVITY_DISPLAY_LIMIT = 30
DISPOSED_STATUSES = {"Sold", "Scrapped"}
SERVICE_STATUSES = {"In Maintenance", "Out of Order"}


def _can(doctype: str, permission: str) -> bool:
    return bool(
        frappe.db.exists("DocType", doctype)
        and frappe.has_permission(doctype, permission)
    )


def _period(from_date=None, to_date=None):
    end = getdate(to_date or nowdate())
    start = getdate(from_date or end.replace(month=1, day=1))
    days = date_diff(end, start)
    if days < 0:
        frappe.throw(frappe._("From date must be before or equal to To date"))
    if days > MAX_PERIOD_DAYS:
        frappe.throw(
            frappe._("Maximum fixed-asset review period in days: {0}").format(
                MAX_PERIOD_DAYS
            )
        )
    return start, end


def _safe_list(
    doctype: str,
    *,
    errors: list[str],
    required: bool = True,
    **kwargs,
) -> list[dict]:
    """Read user-facing evidence only through Frappe's permission query layer."""

    if not _can(doctype, "read"):
        if required:
            errors.append(doctype)
        return []
    try:
        return frappe.get_list(doctype, **kwargs)
    except Exception:
        errors.append(doctype)
        frappe.log_error(
            title=f"bunood_theme: asset workbench {doctype} query stood down"
        )
        return []


def _bounded_list(
    doctype: str,
    *,
    errors: list[str],
    truncated_sources: list[str],
    limit: int,
    required: bool = True,
    **kwargs,
) -> list[dict]:
    rows = _safe_list(
        doctype,
        errors=errors,
        required=required,
        limit_page_length=limit,
        **kwargs,
    )
    if len(rows) >= limit:
        truncated_sources.append(doctype)
        return rows[: limit - 1]
    return rows


def _date_text(value) -> str:
    return str(value or "")


def _is_due(asset: dict, end) -> bool:
    next_date = asset.get("next_depreciation_date")
    if not next_date or not asset.get("calculate_depreciation"):
        return False
    if int(asset.get("docstatus") or 0) != 1:
        return False
    if asset.get("status") in DISPOSED_STATUSES:
        return False
    return getdate(next_date) <= end


def _asset_state(asset: dict, end) -> str:
    status = asset.get("status") or ""
    docstatus = int(asset.get("docstatus") or 0)
    if docstatus == 2 or status == "Cancelled":
        return "cancelled"
    if docstatus == 0 or status == "Draft":
        return "draft-review"
    if status in DISPOSED_STATUSES:
        return "disposed"
    if asset.get("depr_entry_posting_status") == "Failed":
        return "depreciation-failed"
    if _is_due(asset, end):
        return "depreciation-due"
    if status in SERVICE_STATUSES:
        return "service-attention"
    if status == "Fully Depreciated":
        return "fully-depreciated"
    return "in-use"


def _activity_row(doctype: str, row: dict, date_field: str, amount_field=None) -> dict:
    return {
        "doctype": doctype,
        "name": row.get("name"),
        "date": _date_text(row.get(date_field)),
        "docstatus": int(row.get("docstatus") or 0),
        "asset": row.get("asset") or row.get("target_asset") or "",
        "title": row.get("title") or row.get("asset_name") or row.get("purpose") or "",
        "state": row.get("repair_status") or row.get("purpose") or "",
        "amount": flt(row.get(amount_field)) if amount_field else 0,
        "reference_doctype": row.get("reference_doctype") or "",
        "reference_name": row.get("reference_name") or "",
    }


def get_asset_workbench(company: str, from_date=None, to_date=None) -> dict:
    """Return bounded native asset evidence without inferring reconciliation."""

    if not company:
        frappe.throw(frappe._("Company is required"))
    company_doc = frappe.get_doc("Company", company)
    company_doc.check_permission("read")
    start, end = _period(from_date, to_date)
    errors: list[str] = []
    truncated_sources: list[str] = []

    assets = _bounded_list(
        "Asset",
        errors=errors,
        truncated_sources=truncated_sources,
        limit=ASSET_QUERY_LIMIT,
        filters={"company": company},
        fields=[
            "name",
            "asset_name",
            "item_code",
            "item_name",
            "asset_category",
            "status",
            "docstatus",
            "purchase_date",
            "available_for_use_date",
            "calculate_depreciation",
            "next_depreciation_date",
            "depr_entry_posting_status",
            "value_after_depreciation",
            "total_asset_cost",
            "net_purchase_amount",
            "location",
            "custodian",
            "department",
            "disposal_date",
            "maintenance_required",
            "booked_fixed_asset",
            "modified",
        ],
        order_by="modified desc",
    )
    schedules = _bounded_list(
        "Asset Depreciation Schedule",
        errors=errors,
        truncated_sources=truncated_sources,
        limit=AUXILIARY_QUERY_LIMIT,
        filters={"company": company},
        fields=[
            "name",
            "asset",
            "finance_book",
            "depreciation_method",
            "status",
            "docstatus",
            "value_after_depreciation",
            "net_purchase_amount",
            "modified",
        ],
        order_by="modified desc",
    )
    movements = _bounded_list(
        "Asset Movement",
        errors=errors,
        truncated_sources=truncated_sources,
        limit=AUXILIARY_QUERY_LIMIT,
        filters={
            "company": company,
            "transaction_date": [
                "between",
                [f"{start} 00:00:00", f"{end} 23:59:59"],
            ],
        },
        fields=[
            "name",
            "purpose",
            "transaction_date",
            "reference_doctype",
            "reference_name",
            "docstatus",
            "modified",
        ],
        order_by="transaction_date desc, modified desc",
    )
    repairs = _bounded_list(
        "Asset Repair",
        errors=errors,
        truncated_sources=truncated_sources,
        limit=AUXILIARY_QUERY_LIMIT,
        filters={"company": company},
        fields=[
            "name",
            "asset",
            "asset_name",
            "failure_date",
            "completion_date",
            "repair_status",
            "docstatus",
            "capitalize_repair_cost",
            "total_repair_cost",
            "increase_in_asset_life",
            "modified",
        ],
        order_by="failure_date desc, modified desc",
    )
    capitalizations = _bounded_list(
        "Asset Capitalization",
        errors=errors,
        truncated_sources=truncated_sources,
        limit=AUXILIARY_QUERY_LIMIT,
        filters={
            "company": company,
            "posting_date": ["between", [str(start), str(end)]],
        },
        fields=[
            "name",
            "title",
            "target_asset",
            "target_item_code",
            "finance_book",
            "posting_date",
            "total_value",
            "docstatus",
            "modified",
        ],
        order_by="posting_date desc, modified desc",
    )
    adjustments = _bounded_list(
        "Asset Value Adjustment",
        errors=errors,
        truncated_sources=truncated_sources,
        limit=AUXILIARY_QUERY_LIMIT,
        filters={
            "company": company,
            "date": ["between", [str(start), str(end)]],
        },
        fields=[
            "name",
            "asset",
            "finance_book",
            "date",
            "current_asset_value",
            "new_asset_value",
            "difference_amount",
            "journal_entry",
            "docstatus",
            "modified",
        ],
        order_by="date desc, modified desc",
    )

    summary = {
        "observed_assets": len(assets),
        "active_assets": 0,
        "draft_assets": 0,
        "depreciation_due": 0,
        "depreciation_failed": 0,
        "service_attention": 0,
        "disposed_in_period": 0,
        "observed_carrying_value": 0.0,
        "draft_schedules": 0,
        "pending_repairs": 0,
        "lifecycle_drafts": 0,
        "period_activity": len(movements) + len(capitalizations) + len(adjustments),
    }
    presented_assets = []
    for row in assets:
        lifecycle_state = _asset_state(row, end)
        if int(row.get("docstatus") or 0) == 0:
            summary["draft_assets"] += 1
        elif (
            int(row.get("docstatus") or 0) == 1
            and row.get("status") not in DISPOSED_STATUSES
        ):
            summary["active_assets"] += 1
        if lifecycle_state == "depreciation-due":
            summary["depreciation_due"] += 1
        elif lifecycle_state == "depreciation-failed":
            summary["depreciation_failed"] += 1
        if lifecycle_state == "service-attention":
            summary["service_attention"] += 1
        disposal_date = row.get("disposal_date")
        if disposal_date and start <= getdate(disposal_date) <= end:
            summary["disposed_in_period"] += 1
        if int(row.get("docstatus") or 0) == 1 and row.get("status") not in DISPOSED_STATUSES:
            summary["observed_carrying_value"] += flt(
                row.get("value_after_depreciation")
            )
        presented_assets.append(
            {
                "name": row.get("name"),
                "asset_name": row.get("asset_name") or row.get("item_name") or row.get("name"),
                "item_code": row.get("item_code") or "",
                "asset_category": row.get("asset_category") or "",
                "status": row.get("status") or "",
                "state": lifecycle_state,
                "location": row.get("location") or "",
                "custodian": row.get("custodian") or "",
                "department": row.get("department") or "",
                "next_depreciation_date": _date_text(row.get("next_depreciation_date")),
                "depreciation_posting_status": row.get("depr_entry_posting_status") or "",
                "carrying_value": flt(row.get("value_after_depreciation")),
                "total_asset_cost": flt(row.get("total_asset_cost")),
                "purchase_date": _date_text(row.get("purchase_date")),
                "disposal_date": _date_text(row.get("disposal_date")),
            }
        )

    summary["draft_schedules"] = sum(
        1 for row in schedules if int(row.get("docstatus") or 0) == 0
    )
    summary["pending_repairs"] = sum(
        1 for row in repairs if row.get("repair_status") == "Pending"
    )
    summary["lifecycle_drafts"] = sum(
        int(row.get("docstatus") or 0) == 0
        for rows in (movements, repairs, capitalizations, adjustments)
        for row in rows
    )

    priority = {
        "depreciation-failed": 0,
        "depreciation-due": 1,
        "draft-review": 2,
        "service-attention": 3,
        "in-use": 4,
        "fully-depreciated": 5,
        "disposed": 6,
        "cancelled": 7,
    }
    presented_assets.sort(
        key=lambda row: (priority.get(row["state"], 8), row["asset_name"], row["name"])
    )

    activity = []
    activity.extend(
        _activity_row("Asset Movement", row, "transaction_date") for row in movements
    )
    activity.extend(
        _activity_row("Asset Repair", row, "failure_date", "total_repair_cost")
        for row in repairs
        if row.get("repair_status") == "Pending"
        or (row.get("failure_date") and start <= getdate(row.get("failure_date")) <= end)
    )
    activity.extend(
        _activity_row("Asset Capitalization", row, "posting_date", "total_value")
        for row in capitalizations
    )
    activity.extend(
        _activity_row("Asset Value Adjustment", row, "date", "difference_amount")
        for row in adjustments
    )
    activity.sort(key=lambda row: row["date"], reverse=True)

    if errors or truncated_sources:
        state = "incomplete-evidence"
    elif summary["depreciation_failed"] or summary["depreciation_due"] or summary["pending_repairs"]:
        state = "attention-needed"
    elif summary["draft_assets"] or summary["draft_schedules"] or summary["lifecycle_drafts"]:
        state = "review-required"
    else:
        state = "no-exceptions-observed"

    return {
        "company": company,
        "currency": company_doc.get("default_currency"),
        "finance_book": company_doc.get("default_finance_book"),
        "from_date": str(start),
        "to_date": str(end),
        "generated_at": str(now_datetime()),
        "state": state,
        "query_errors": list(dict.fromkeys(errors)),
        "truncated_sources": list(dict.fromkeys(truncated_sources)),
        "evidence_truncated": bool(truncated_sources),
        "summary": summary,
        "capabilities": {
            "can_read_asset": _can("Asset", "read"),
            "can_create_asset": _can("Asset", "create"),
            "can_create_movement": _can("Asset Movement", "create"),
            "can_create_repair": _can("Asset Repair", "create"),
            "can_create_capitalization": _can("Asset Capitalization", "create"),
            "can_create_adjustment": _can("Asset Value Adjustment", "create"),
            "can_read_gl": _can("GL Entry", "read"),
        },
        "assets": presented_assets[:ASSET_DISPLAY_LIMIT],
        "asset_display_truncated": len(presented_assets) > ASSET_DISPLAY_LIMIT,
        "schedules": schedules[:ACTIVITY_DISPLAY_LIMIT],
        "schedule_display_truncated": len(schedules) > ACTIVITY_DISPLAY_LIMIT,
        "activity": activity[:ACTIVITY_DISPLAY_LIMIT],
        "activity_display_truncated": len(activity) > ACTIVITY_DISPLAY_LIMIT,
    }
